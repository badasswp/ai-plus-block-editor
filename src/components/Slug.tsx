import { __ } from '@wordpress/i18n';
import { check } from '@wordpress/icons';
import { select, dispatch } from '@wordpress/data';
import { Button, TextControl, Icon } from '@wordpress/components';
import { useState, useEffect } from '@wordpress/element';
import apiFetch from '@wordpress/api-fetch';

import Toast from '../components/Toast';

/**
 * Slug.
 *
 * This Component returns the Slug
 * label and button.
 *
 * @since 1.1.0
 *
 * @returns {JSX.Element}
 */
const Slug = (): JSX.Element => {
  const [ slug, setSlug ] = useState( '' );
  const [ isLoading, setIsLoading ] = useState( false );
  const { editPost } = dispatch( 'core/editor' ) as any;
  const {
    getCurrentPostId,
    getEditedPostContent,
    getEditedPostAttribute
  } = select( 'core/editor' );

  const content = getEditedPostContent();

  useEffect( () => {
    setSlug( getEditedPostAttribute( 'meta' )['apbe_slug'] );
  }, [] )

  /**
   * This function fires when the user clicks
   * the `Generate` button.
   *
   * @since 1.1.0
   *
   * @returns { void }
   */
  const handleClick = async (): Promise<void> => {
    setIsLoading( true );

    const response = await apiFetch(
      {
        path: '/ai-plus-block-editor/v1/sidebar',
        method: 'POST',
        data: {
          id: getCurrentPostId(),
          text: content.text || content,
          feature: 'slug'
        },
      }
    );

    const { data } = response as any;
    const slashedSlug = ( '/' != data.charAt(0) ) ? `/${data}` : data;

    let limit = 1;

    const showAnimatedAiText = setInterval( () => {
      if ( limit === slashedSlug.length ) {
        clearInterval( showAnimatedAiText );
      }
      setSlug( slashedSlug.substring( 0, limit ) );
      limit++;
    }, 5 )

    setIsLoading( false );
  }

  /**
   * This function fires when the user selects
   * the AI generated result.
   *
   * @since 1.1.0
   *
   * @returns { void }
   */
  const handleSelection = (): void => {
    editPost( { slug } );
    editPost( { meta: { apbe_slug: slug } } );
  }

  return (
    <>
      <p><strong>{ __( 'Slug', 'ai-plus-block-editor' ) }</strong></p>
      <TextControl
        __nextHasNoMarginBottom
        placeholder="/your-article-slug"
        value={ slug }
        onChange={ () => { } }
      />
      <div className="apbe-button-group">
        <Button
          variant="primary"
          onClick={ handleClick }
        >
          { __( 'Generate', 'ai-plus-block-editor' ) }
        </Button>
        <Button
          variant="secondary"
          onClick={ handleSelection }
        >
          <Icon icon={ check } />
        </Button>
      </div>
      <Toast
        message={ __( 'AI is generating text, please hold on for a bit...' ) }
        isLoading={ isLoading }
      />
    </>
  )
}

export default Slug;
