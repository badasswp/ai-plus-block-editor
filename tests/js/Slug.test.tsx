import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';

import Slug from '../../src/components/Slug';

jest.mock( '@wordpress/i18n', () => ( {
	__: jest.fn( ( arg ) => arg ),
} ) );

jest.mock( '@wordpress/components', () => ( {
	Button: jest.fn( ( { children, variant, onClick } ) => {
		return (
			<>
				<button className={ variant } onClick={ onClick }>
					{ children }
				</button>
			</>
		);
	} ),
	Icon: jest.fn( () => {
		return <>Icon</>;
	} ),
	TextareaControl: jest.fn(
		( { rows, value, onChange, __nextHasNoMarginBottom = true } ) => {
			return (
				__nextHasNoMarginBottom && (
					<>
						<textarea
							rows={ rows }
							onChange={ onChange }
							value={ value }
						/>
					</>
				)
			);
		}
	),
	TextControl: jest.fn(
		( {
			placeholder,
			value,
			onChange,
			__nextHasNoMarginBottom = true,
		} ) => {
			return (
				__nextHasNoMarginBottom && (
					<>
						<input
							placeholder={ placeholder }
							onChange={ onChange }
							value={ value }
						/>
					</>
				)
			);
		}
	),
} ) );

jest.mock( '@wordpress/data', () => ( {
	dispatch: jest.fn( ( store ) =>
		store === 'core/editor'
			? {
					editPost: jest.fn(),
			  }
			: {}
	),
	select: jest.fn( ( store ) =>
		store === 'core/editor'
			? {
					savePost: jest.fn(),
					getCurrentPostId: jest.fn(),
					getEditedPostContent: jest.fn(),
					getEditedPostAttribute: jest.fn( ( attribute ) =>
						attribute === 'meta'
							? {
									apbe_slug: 'ai-generated-slug',
							  }
							: {}
					),
			  }
			: {}
	),
} ) );

describe( 'Slug', () => {
	it( 'renders the Slug text input and 2 buttons', () => {
		const { container } = render( <Slug /> );

		// Expect Component to look like so:
		expect( container.innerHTML ).toBe(
			`<p><strong>Slug</strong></p><input placeholder="your-article-slug" value="ai-generated-slug"><div class="apbe-button-group"><button class="primary">Generate</button><button class="secondary">Icon</button></div>`
		);

		// Assert the Generate button is displayed.
		const generateButton = screen.getByText( 'Generate' );
		expect( generateButton ).toHaveClass( 'primary' );
		expect( generateButton ).toBeInTheDocument();
		expect( generateButton ).toBeInstanceOf( HTMLButtonElement );

		// Assert the Select button is displayed.
		const selectButton = screen.getByText( 'Icon' );
		expect( selectButton ).toHaveClass( 'secondary' );
		expect( selectButton ).toBeInTheDocument();
		expect( selectButton ).toBeInstanceOf( HTMLButtonElement );
	} );
} );
