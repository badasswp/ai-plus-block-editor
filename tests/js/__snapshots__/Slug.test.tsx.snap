// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slug renders Slug component 1`] = `
<div>
  <p>
    <strong>
      Slug
    </strong>
  </p>
  <input
    placeholder="your-article-slug"
    value="ai-generated-slug"
  />
  <div
    class="apbe-button-group"
  >
    <button
      class="primary"
    >
      Generate
    </button>
    <button
      class="secondary"
    >
      Icon
    </button>
  </div>
</div>
`;
